src/schema/posts.ts

```diff
@@ -1371,6 +1391,58 @@ export const resolvers: IResolvers<any, Context> = {
 
       return { _: true };
     },
+    swapPinnedPosts: async (
+      _,
+      { id, swapWithId }: SwapPinnedPostArgs,
+      ctx: Context,
+    ): Promise<GQLEmptyResponse> => {
+      await ctx.con.transaction(async (manager) => {
+        const repo = manager.getRepository(Post);
+        const post = await repo.findOneBy({ id });
+        const swapWithPost = await repo.findOneBy({
+          id: swapWithId,
+        });
+
+        await ensureSourcePermissions(
+          ctx,
+          post.sourceId,
+          SourcePermissions.PostPin,
+        );
+
+        if (!post.pinnedAt || !swapWithPost.pinnedAt) {
+          throw new ValidationError('Posts must be pinned first');
+        }
+
+        const isNextPost = swapWithPost.pinnedAt > post.pinnedAt;
+        const swapPinnedTime = new Date(
+          swapWithPost.pinnedAt.getTime() + 1000 * (isNextPost ? 1 : -1),
+        );
+        const operation = isNextPost
+          ? `+ interval '1 second'`
+          : `- interval '1 second'`;
+        const comparison = isNextPost ? '>=' : '<=';
+
+        await manager.query(
+          `
+          UPDATE post
+          SET "pinnedAt" = "pinnedAt" ${operation}
+          WHERE "pinnedAt" IS NOT NULL
+            AND "pinnedAt" ${comparison} $1
+            AND "sourceId" = $2
+        `,
+          [swapPinnedTime, post.sourceId],
+        );
```

Comment:
Please use `QueryBuilder`

Reply:
@capJavert for non-standard queries I usually prefer using as much SQL as possible vs ORM functions as it is much more readable. That said, I'm fine with the alternative if you prefer. For context, it would look something like:

```typescript
await manager.createQueryBuilder()
    .update(Post)
    .set({ 
        pinnedAt: () => `"pinnedAt" ${operation}`
    })
    .where('"pinnedAt" IS NOT NULL')
    .andWhere(`"pinnedAt" ${comparison} :swapPinnedTime`, { swapPinnedTime })
    .andWhere('"sourceId" = :sourceId', { sourceId: post.sourceId })
    .execute();
```

Reply:
ok, sure, I'll update then. I may not be very familiar with TypeORM, but I am pretty sure that making use of parameterized arguments (`$1`, `$2`, etc.)  achieves the same result of sanitizing arguments and preventing SQL injection, so the current approach should be safe as well?

Reply:
Honestly not sure as well but it differs from ORM to ORM, that is why I think query builder is again safer and better practice.

Reply:
stuff as `${comparison}` are most dangerous, yes its not user provided now but it could become by mistake, with query builder you can even add different values conditionally with `andWhere`.